apps/consulting/posts/scaling-python.md

```diff
@@ -0,0 +1,214 @@
+---
+title: 'Scaling Python'
+published: March 28, 2023
+author: andrew-fulton
+description: 'In this post, we walk you through how Quansight helped a banking client through this process of scaling in a real life scenario.'
+category: [Scalable Computing, Optimization, Data Engineering]
+featuredImage:
+  src: /posts/scaling-python/header.png
+  alt: 'summing dataframes visualization'
+hero:
+  imageSrc: /posts/scaling-python/header.png
+  imageAlt: 'summing dataframes visualization'
+---
+
+<base target="_blank" />
+
+This post is based off of a talk given at PyData NYC 2022. You can watch the talk here: [Scaling Python: Bank Edition](https://www.youtube.com/watch?v=tbqgsM6iGng).
+
+Often scaling a distributed computation is easy until it isn't. There are currently several tools that will get you off the ground quickly, but when the size of your data increases and you are required to scale further, the complexity needed for a solution often increases considerably as well. At this point algorithm and infrastructure dependent requirements will likely preclude experimentation and problem solving. In this post, we walk you through how Quansight helped a banking client through this process of scaling in a real life scenario.
+
+## Problem Overview
+
+To start, we were tasked with deploying a large scale data processing pipeline for running valuation adjustment models using open source tools from the PyData ecosystem. We were given 140,000 simulation files, each around 140MB, totaling about 20 TB. The data was stored in flat parquet files with columns “path”, “Date”, and “value”. Each file contained 50,000 simulations 120 dates into the future (6 million rows) (~140MB/file). Each file belonged to a particular file group, with each group ranging from 2 to 18,000 files. Our goal was to sum the values in the files by path, date within each file group as quickly as possible to return results in as close to real time as possible. To complicate the problem, we had to be able to accomplish this while accounting for potentially limited compute resources.
```

Comment:
"days" makes more sense to me than "dates", but if you meant 'dates' for a particular reason that's fine to keep.

You already said "each around 140MB" earlier in the paragraph, so I figure no need to say it again?

The "sum the values in the files by path, date" part is confusing to me and will need to be rephrased. I don't understand clearly enough what you're going for to offer a suggestion. Are you saying that the sums that needed to be carried out were effectively across the compound key `(path, date)`?

```suggestion
In this project, we were tasked with deploying a large scale data processing pipeline for running valuation adjustment models using open source tools from the PyData ecosystem. We were given 140,000 simulation files, each around 140MB, totaling about 20 TB. The data was stored in flat Parquet files with columns “path”, “Date”, and “value”. Each file contained 50,000 simulations per day, spanning 120 days into the future (6 million rows). Each file belonged to a particular file group, with each group ranging from 2 to 18,000 files. Our goal was to sum the values in the files by path, date within each file group as quickly as possible to return results in as close to real time as possible. To complicate the problem, we had to be able to accomplish this while accounting for potentially limited compute resources.
```

Reply:
The problem with days, is that it makes it sounds consecutive. The dates were often 1 per quarter, I'm not sure if they could be different periods than that though.

Reply:
@bskinn, any thoughts on this? ^

Keep or change to something else?

Reply:
Oof. Yeah, even after that explanation I still don't really get how the data is laid out.

That's kind of my concern, though, is that right now it gives enough information to get a reader pretty interested in what the data is like, but not enough to fully understand it.

I think your suggestion might be on the right track, not 100% sure yet.

---

So, terminology. When I look at that figure you posted in the comment, I see a number of traces on the graph. E.g., a red trace ("Original"), a purple trace, an orange trace, etc.

Is it the case that there were 50,000 traces per file, and each trace had 120 points?

And, is what I'm just now calling a "trace" the same thing as the data calls a "path"?

Reply:
Yeah, that's exactly right. The image below this paragraph, has a picture that shows the dataframes with their data being summed. We were just taking value of a path on a date and summing it with the other values for that same date and path from other files.

Reply:
Ok, great. 

How about we leave this suggestion un-merged, I propose a different phrasing as part of my review pass, and then I make sure to flag it so that you can evaluate its correctness?